syntax = "proto3";

package yscb_example.messages;

import "google/protobuf/any.proto";

message Wrapper { 
    string request_id = 1;
    google.protobuf.Any message = 2;
}

message State {
    map<string,string> fields = 1;
    int32 balance = 2;
}

// Outside interface
message Insert {
    string id = 1;
    State state = 2;
}

message Read {
    string id = 1;
}

message Update {
    string id = 1;
    map<string, string> updates = 2;
}

message DeleteAndTransferAll {
    string id = 1;
    string incoming_id = 2;
}

message Transfer {
    string outgoing_id = 1;
    string incoming_id = 2;
    int32 amount = 3;
}

message Response {
    string request_id = 1;
    int32 status_code = 2;
    google.protobuf.Any message = 3;
}

// Internal messages
message AddCredit { 
    int32 amount = 1;
}

message SubtractCredit {
    int32 amount = 1;
}

message Delete {

}